### DISCUSSION TITLE
Annotations lost when using statement.froms

### DISCUSSION THREAD

**Discussion by mjpieters** [sqlalchemy/sqlalchemy#6807] [Usage Questions]
We make use of `with_loader_criteria()`, which makes use of the `parententity` annotations normally attached to tables pointing back to their mapper (the ORMSelectCompileState context class retrieves these to match against the global criteria provided by the loader).

However, we found that using `Query.statement.froms` means we lose those annotations. Is this intentional?

We have a `rowcounts` property on our `Query` subclass to transform an existing query into an efficient rowcount query:

```py
    @property
    def rowcount(self):
        """Rowcount for query

        This is more efficient than using .count() (which uses a subselect)

        """
        query = (
            # preserve FROM when changing columns
            self.statement.select_from(*self.statement.froms)
            # replace selected columns with COUNT(*)
            .with_only_columns(sql.func.count())
            # drop ORDER BY, counting doesn't care about ordering
            .order_by(None)
        )
        return self.session.execute(query).scalar()
```

**To Reproduce**
MRE:

```py
from sqlalchemy import Column, Integer, String, create_engine, event, func
from sqlalchemy.orm import Session, declarative_base, sessionmaker, with_loader_criteria


@event.listens_for(Session, "do_orm_execute")
def answers_to_life_the_universe_and_everything_only(execute_state):
    if execute_state.is_column_load:
        return

    execute_state.statement = execute_state.statement.options(
        with_loader_criteria(QA, lambda cls: cls.answer == "42", include_aliases=True)
    )

Base = declarative_base()


class QA(Base):
    __tablename__ = "qa"
    id = Column(Integer, primary_key=True)
    question = Column(String)
    answer = Column(String)

    def __str__(self):
        return f"«{self.question}»\n-- ‘{self.answer}’"


engine = create_engine("sqlite://")
session = sessionmaker(bind=engine)()
Base.metadata.create_all(bind=engine)

session.add_all(
    [
        QA(question="How many roads must a man walk down?", answer="42"),
        QA(question="But look, you found the notice, didn’t you?", answer="Yes"),
    ]
)
session.commit()

all_qa = session.query(QA)
rowcount = all_qa.statement.select_from(*all_qa.statement.froms).with_only_columns(
    func.count()
)

print(*all_qa.all())
# prints 2, should print 1
print("Total questions:", session.execute(rowcount).scalar())
```

The above outputs:

```none
«How many roads must a man walk down?»
-- ‘42’
Total questions: 2
```

but the expected output is `Total questions: 1`.

**Versions.**
 - OS: macos 11.5
 - Python: 3.9.6
 - SQLAlchemy: 1.4.20
 - Database: SQLite, PostgreSQL
 - DBAPI: sqlite3, psycopg2

**Additional context**
I work around this issue by using the entities provided by `statement.get_children()`:

```py
    @property
    def rowcount(self):
        """Rowcount for query

        This is more efficient than using .count() (which uses a subselect)

        """
        # use the original entities, which have annotations that make
        # it possible to apply loader criteria to the new query. These
        # annotations are lost when using self.statement.froms
        orig_entities = {ent: ent for ent in self.statement.get_children()}
        # annotations are not part of the hash key, so entity will match
        froms = (orig_entities.get(entity, entity) for entity in self.statement.froms)

        query = (
            # preserve FROM when changing columns
            self.statement.select_from(*froms)
            # replace selected columns with COUNT(*)
            .with_only_columns(sql.func.count())
            # drop ORDER BY, counting doesn't care about ordering
            .order_by(None)
        )
        return self.session.execute(query).scalar()
```

**Comment by zzzeek** [comment]
OK so I converted to discussion here as I saw things with legacy Query and statement and that confused me into thinking you were using pure Core statements coming from Query and expecting ORM-ish things to keep remaining.   but then I saw the 2.0 execute at the end, so we can restate this as a pure 2.0 problem so that I dont have to think there are legacy patterns being mixed.

what you are doing more succinctly is this:

```py
all_qa = select(QA)

rowcount = all_qa.select_from(*all_qa.froms).with_only_columns(func.count())
```

when what the docs say you would do is this:

```py
all_qa = select(QA)

rowcount = all_qa.select_from(QA).with_only_columns(func.count())

```

now what is bad here about SQLAlchemy is the lazy implementation of with_only_columns() that doesn't have any logic to maintain the FROM list that was established initially.

so somehow you found "select.froms" and are reusing that (nobody has ever used that before, but yes now that it's ORM of course people will), and right now the way the ORM gets that for you is completely inefficient and goes way beyond how it should be getting you those and is returning them from the core SELECT that's supposed to be only generated during compilation that has no annotations, so it's completely wrong.

in short, github gives me no way to revert this to an issue so i will delete this and start all over again.

**Reply by zzzeek** [reply]
it's a big behavioral change and should have been made in 1.4 betas.    I'm just going to do it, assuming tests dont uncover some important use case, because i dont think .froms was really useful in this way in the 1.3 series.

**Reply by zzzeek** [reply]
OK, use case found:

```
        stmt = select(User).join(User.addresses)

        assert stmt.froms[0].compare(users.join(addresses))
```


full ORM compilation must proceed :(

I think you need to just have a new method here, because calling stmt = stmt.select_from(stmt.froms) is wildly inefficient.

**Reply by CaselIT** [reply]
maybe we could have two new properties and deprecate `froms`?
One called something like `all_froms` thats the same a the current `froms` and the other called something like `main_froms` or `initial_froms` or something similar that gives back the froms provided by the user without doing any compilation

**Reply by zzzeek** [reply]
this is in #6808.    See https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2973 and modify your program as:

```py
all_qa = session.query(QA)
rowcount = all_qa.statement.with_only_columns(
    func.count(), maintain_column_froms=True
)
```

**Reply by zzzeek** [reply]
> maybe we could have two new properties and deprecate `froms`?
> One called something like `all_froms` thats the same a the current `froms` and the other called something like `main_froms` or `initial_froms` or something similar that gives back the froms provided by the user without doing any compilation

yes that's sort of what I did without deprecating .froms yet, I can see a better name for ".froms" called ".final_froms", so we can do that too.

the gerrit includes ".columns_clause_froms" as the new attribute.  this is the main thing people need to have and really it's just the flag on with_only_columns().  Anddd....it maybe should just be another method "only_include_columns()" or something like that but these names are all vague in terms of this aspect of it.

**Reply by mjpieters** [reply]
Yes, I was using `.froms` to work around the fact that `with_only_columns()` dropped all froms, and I kind-a noticed that it was going through the compilation machinery to do this, but I had not found a better way. :-)

I'll move this over to the new report.

**Comment by decaz** [comment]
Let me add the following remark, please. At version 1.3 there was `statement._froms` list attribute which I could use to know what **tables** were affected by `sqlalchemy.sql.Select` statement. It is used within my test suites to intercept queries to specific tables. Now, at version 1.4 I've found the following hack to detect these tables:
```python
affected_tables = []
for query_from in statement.froms:
    affected_tables += query_from._from_objects
```
Can you tell me please is there any correct way to get the list of selected tables as it was at version 1.3? I guess this is relevant to this discussion.

**Reply by zzzeek** [reply]
the best way to get froms is to visit them, see https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/sql/util.py#L237 for a comprehensive way to do this taking into account different kinds of FROM clauses.

**Reply by decaz** [reply]
@zzzeek thank you!
